### PR TITLE
[IMP] mail: show yellow notification instead of send message in chatter when joined and unsubscribed channels

### PR DESCRIPTION
**Pull request by dep-odoo**
Task: https://www.odoo.com/web#id=1839849&view_type=form&model=project.task&action=333&active_id=965&menu_id=4720
Pad: https://pad.odoo.com/p/r.46c4d20bd4d7e6605bd68ef5b0eef95d

### CODE REVIEW

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 1]
```diff
@@ -169,6 +169,22 @@ MailManager.include({
...
+     * Called when receiving a join and left channel notification:
+     *
+     * @private
+     * @param {Object} channelData
+     * @param {string} channelData.channel_partner_id
+     * @param {string} channelData.channel_partner_name
+     * @param {string} channelData.name
+     */
+    _handlePartnerJoinLeftNotification: function (channelData) {
+        var username = this.getSession().partner_id == channelData.channel_partner_id ?
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: String concatenation on translatable string is a bad idea.

Here is a suggested way to make translatable string:

```
var title;
var content;
var isOnMyself = session.partner_id === channelData.channel_partner_id;
if (channelData.info === 'joined') {
    title = _t("Joined");
    if (isOnMyself) {
        content = _.str.sprintf(
            _t("You have joined channel <b>%s</b>."),
            channelData.name
        );
    } else {
        content = _.str.sprintf(
            _t("%s has joined channel <b>%s</b>."),
            channelData.channel_partner_name,
            channelData.name
        );
    }
} else if (channelData.info === 'left') {
    title = _t("Left");
    if (isOnMyself) {
        content = _.str.sprintf(
            _t("You have left channel <b>%s</b>."),
            channelData.name
        );
    } else {
        content = _.str.sprintf(
            _t("%s has left channel <b>%s</b>."),
            channelData.channel_partner_name,
            channelData.name
        );
    }
}
```

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 2]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
...
+     * @param {string} channelData.channel_partner_name
+     * @param {string} channelData.name
+     */
+    _handlePartnerJoinNotification: function (channelData) {
+        var content;
+        var isOnMyself = session.partner_id === channelData.channel_partner_id;
+        if (isOnMyself) {
+            content = _.str.sprintf(
+                _t("You have joined channel : <b>%s</b>"),
+                channelData.name
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: should escape `channelData.name`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 3]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
...
+            );
+        } else {
+            content = _.str.sprintf(
+                _t("%s has joined channel : <b>%s</b>"),
+                channelData.channel_partner_name,
+                channelData.name
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: should escape `channelData.name`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 4]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
...
+        var isOnMyself = session.partner_id === channelData.channel_partner_id;
+        if (isOnMyself) {
+            content = _.str.sprintf(
+                _t("You have joined channel : <b>%s</b>"),
+                channelData.name
+            );
+        } else {
+            content = _.str.sprintf(
+                _t("%s has joined channel : <b>%s</b>"),
+                channelData.channel_partner_name,
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: should escape `channelData.channel_partner_name`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 5]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
             }
         }
     },
+    /**
+     * Called when receiving a join and left channel notification:
+     *
+     * @private
+     * @param {Object} channelData
+     * @param {string} channelData.channel_partner_id
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: `{integer}`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 6]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
...
+     * @param {string} channelData.channel_partner_name
+     * @param {string} channelData.name
+     */
+    _handlePartnerJoinNotification: function (channelData) {
+        var content;
+        var isOnMyself = session.partner_id === channelData.channel_partner_id;
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]:
```js
if (session.partner_id === channelData.channel_partner_id) {
   ...
}
```

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 7]
```diff
@@ -169,6 +169,32 @@ MailManager.include({
...
+    /**
+     * Called when receiving a join and left channel notification:
+     *
+     * @private
+     * @param {Object} channelData
+     * @param {string} channelData.channel_partner_id
+     * @param {string} channelData.channel_partner_name
+     * @param {string} channelData.name
+     */
+    _handlePartnerJoinNotification: function (channelData) {
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: this method should be placed after `_handlePartnerChannelNotification ` (alphabetical order)

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 8]
```diff
@@ -429,24 +457,38 @@ MailManager.include({
      * @private
      * @param {Object} data
      * @param {Object} data.id ID of the unsubscribed channel
+     * @param {String} data.channel_partner_id
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: `{integer}`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 9]
```diff
@@ -429,24 +457,38 @@ MailManager.include({
...
+                if (isOnMyself) {
+                    message = _.str.sprintf(
+                        _t("You unsubscribed from <b>%s</b>."),
+                        channel.getName()
+                    );
+                } else {
+                    message = _.str.sprintf(
+                        _t("%s has left channel : <b>%s</b>"),
+                        data.channel_partner_name,
+                        data.name
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: you should escape `data.name`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 10]
```diff
@@ -429,24 +457,38 @@ MailManager.include({
...
+                isOnMyself = session.partner_id === data.channel_partner_id;
+                if (isOnMyself) {
+                    message = _.str.sprintf(
+                        _t("You unsubscribed from <b>%s</b>."),
+                        channel.getName()
+                    );
+                } else {
+                    message = _.str.sprintf(
+                        _t("%s has left channel : <b>%s</b>"),
+                        data.channel_partner_name,
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: you should escape `data.channel_partner_name`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 11]
```diff
@@ -429,24 +457,38 @@ MailManager.include({
      * @private
      * @param {Object} data
      * @param {Object} data.id ID of the unsubscribed channel
+     * @param {String} data.channel_partner_id
+     * @param {String} data.name
      */
     _handlePartnerUnsubscribeNotification: function (data) {
         var channel = this.getChannel(data.id);
+        var isOnMyself = true;
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: `var isOnMyself = session.partner_id === data.channel_partner_id;` (evaluate once)

**Comment by dep-odoo** [comment]
@alexkuhn changes are completed.

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 12]
```diff
@@ -428,25 +455,38 @@ MailManager.include({
...
             var message;
             if (_.contains(['public', 'private'], channel.getType())) {
-                message = _.str.sprintf(
-                    _t("You unsubscribed from <b>%s</b>."),
-                    channel.getName()
-                );
+                if (isOnMyself) {
+                    message = _.str.sprintf(
+                        _t("You unsubscribed from <b>%s</b>."),
+                        channel.getName()
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: `_.escape(channel.getName())`

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 13]
```diff
@@ -428,25 +455,38 @@ MailManager.include({
...
+                    message = _.str.sprintf(
+                        _t("%s has left channel : <b>%s</b>"),
+                        _.escape(data.channel_partner_name),
+                        _.escape(data.name)
+                    );
+                }
             } else {
                 message = _.str.sprintf(
                     _t("You unpinned your conversation with <b>%s</b>."),
                     channel.getName()
```
Comment (alexkuhn) [inline addons/mail/static/src/js/services/mail_notification_manager.js]: `_.escape(channel.getName())`

**Comment by dep-odoo** [comment]
@alexkuhn  changes are completed.

**Comment by Xavier-Do** [comment]
I'm sorry but this is not a correct solution for me:
From an implementation point of view, to many messages between server and client. We should send notification on subscribe, not when the client notifies that he received a notification saying that he joined a channel.
But...
The real problem we have now is that js will notify the server that the user joined a new channel, even if in reality the server already knows that or if the user was already in the channel. 
->Server notifies client and notification is cached by _handleChannelMessageNotification
I'm trying to find all the possible message that can be send and catched by this handler. Since we dont have a precise message to grep (fallback case) 
->Js dont have the channel in memory for some reasons: call channel_join_and_get_info
Most of the time this road is called while the user is already in the channel. A first possible fix is to check that user is not in the channel_partner_ids. A little tweak on action_follow to return the result of the operation could be enough: only send the message when action_follow did something. 
But as a reminder, we should also avoid to call channel_join_and_get_info for no reason as we are doing now. I will see that with @alexkuhn next week.  

For this task, wait a little until we are sure of what we want to do.

**Comment by rdeodoo** [comment]
@Xavier-Do Task cancelled a few months ago, apparently the bug was fixed meanwhile but I am not sure given your last message on the task if this one should be closed or not

**Pull request closed without merging**